kahypar/partition/initial_partition.h

```diff
@@ -155,64 +137,53 @@ static inline Context createContext(const Hypergraph& hg,
 }
 
 
-static inline void partition(Hypergraph& hg, const Context& context) {
+static inline void partition(Hypergraph& hg,
+                             const Context& context) {
   auto extracted_init_hypergraph = ds::reindex(hg);
+  Hypergraph& init_hg = *extracted_init_hypergraph.first;
   std::vector<HypernodeID> mapping(std::move(extracted_init_hypergraph.second));
-  double init_alpha = context.initial_partitioning.init_alpha;
-  double best_imbalance = std::numeric_limits<double>::max();
-  std::vector<PartitionID> best_balanced_partition(
-    extracted_init_hypergraph.first->initialNumNodes(), 0);
 
   double fixed_vertex_subgraph_imbalance = 0.0;
   if (hg.numFixedVertices() > 0) {
     fixed_vertex_subgraph_imbalance = metrics::imbalanceFixedVertices(hg, context.partition.k);
   }
 
-  do {
-    extracted_init_hypergraph.first->resetPartitioning();
-    // we do not want to use the community structure used during coarsening in initial partitioning
-    extracted_init_hypergraph.first->resetCommunities();
-    Context init_context = createContext(*extracted_init_hypergraph.first, context, init_alpha);
-
-    if (context.initial_partitioning.verbose_output) {
-      LOG << "Calling Initial Partitioner:" << context.initial_partitioning.technique
-          << context.initial_partitioning.mode
-          << context.initial_partitioning.algo
-          << "(k=" << init_context.initial_partitioning.k << ", epsilon="
-          << init_context.partition.epsilon << ")";
-    }
-    if ((context.initial_partitioning.technique == InitialPartitioningTechnique::flat &&
-         context.initial_partitioning.mode == Mode::direct_kway) ||
+  // we do not want to use the community structure used during coarsening in initial partitioning
+  init_hg.resetCommunities();
+  Context init_context = createContext(init_hg, context);
+
+  if (context.initial_partitioning.verbose_output) {
+    LOG << "Calling Initial Partitioner:" << context.initial_partitioning.technique
+        << context.initial_partitioning.mode
+        << context.initial_partitioning.algo
+        << "(k=" << init_context.initial_partitioning.k << ", epsilon="
+        << init_context.partition.epsilon << ")";
+  }
+  if ((context.initial_partitioning.technique == InitialPartitioningTechnique::flat &&
+        context.initial_partitioning.mode == Mode::direct_kway) ||
+      fixed_vertex_subgraph_imbalance > context.partition.epsilon) {
+    // NOTE: If the fixed vertex subgraph is imbalanced, we cannot guarantee
+    //       a balanced initial partition with recursive bisection. Therefore,
+    //       we switch to direct k-way flat mode in that case.
+    if (!(context.initial_partitioning.technique == InitialPartitioningTechnique::flat &&
```

Comment:
I have adjusted the check here such that the warning ("Fixed vertex subgraph is imbalanced") may only be triggered before initial partitioning. Previously, the warning sometimes was (rather inappropriately) triggered by the fixed vertices of the prepacking algorithm.

However, I guess we could the remove the `fixed_vertex_subgraph_imbalance > context.partition.epsilon` check, the warning and the switching to flat mode altogether, as with the inclusion of this PR there is no more reason to believe that flat partitioning provides better balance than recursive bisection.

Reply:
Should we do some small experiments to verify that or can we already say that based on the experiments we did so far?

Reply:
About the balance guarantees of the recursive bisection algorithm I don't have much doubt. After all they are the main point of the upcoming paper, including a formal proof. I.e. the RB algorithm can find a balanced partition if the used load balancing algorithm can find one. Flat partitioning can hardly do better, except of course implementing a better load balancing algorithm.